pkgs/dart_tooling_mcp_server/lib/src/mixins/dtd.dart

```diff
@@ -186,6 +210,21 @@ base mixin DartToolingDaemonSupport on LoggingSupport, ToolsSupport {
       }
     });
     dtd.streamListen('Service');
+
+    dtd.onEvent('Editor').listen((e) async {
+      log(LoggingLevel.debug, e.toString());
+      switch (e.kind) {
+        case 'debugSessionStarted':
+        case 'debugSessionChanged':
+          await updateActiveVmServices();
+        case 'debugSessionStopped':
+          await activeVmServices
+              .remove((e.data['debugSession'] as DebugSession).id)
+              ?.then((service) => service.dispose());
```

Comment:
we are already doing this above when the vm service shuts down, do we need to do it here too? will calling dispose twice be a problem? Also, we'd want to avoid calling this here if if doesn't allow us to remove the resource in the onDone handler above

Reply:
It is safe to call `dispose` multiple times.

This also doesn't affect our ability to dispose of the resource (but will trigger it by triggering the `onDone` future).

Really, this is just defensive, in case somehow the VM service remained active but the debug session was completed. We will shut it down if either DTD tells us to, or the VM service connection closes.

Reply:
I did remove the (should be) unnecessary `dispose` call form the `onDone` callback though, that only gets completed when the VM service is disposed anyways.